Delete VR before PVC deletion

ShyamsundarR opened the pull request:
We had a leak in VR if the reconcile was
interrupted, this is fixed by deleting the VR before
marking the PVC as no longer protected.

This helps with finalizer updates, when there are races
to finalize VRG due to PVC events, and the state fails
in between.

We also ensure VR is in the desired state before deleting
the same, and finalizing the VRG, to leave the system in
an overall desired and consistent state.

Also, corrects S3 object deletion to delete all PVs inside the
PVC bucket.

Fixes #52

Signed-off-by: Shyamsundar Ranganathan <srangana@redhat.com>

BenamarMk commented on controllers/volumereplicationgroup_controller.go:
```diff
@@ -665,15 +661,17 @@ func (v *VRGInstance) reconcileVRForDeletion(pvc *corev1.PersistentVolumeClaim,
 
-	if err := v.preparePVCForVRDeletion(pvc, log); err != nil {
-		log.Info("Requeuing due to failure in preparing PersistentVolumeClaim for VolumeReplication deletion",
+	// Deleting VR first may end-up recreating the VR if reconcile for this PVC is interrupted, but that is better than
```
Question: If the reconcile for this PVC is interrupted, why do you think it will end up recreating the VR? isn't it protected by [this line ](https://github.com/RamenDR/ramen/blob/400fddb4c913a1e9f53af02d51a71fe5fa51ce53/controllers/volumereplicationgroup_controller.go#L377). IOW, the VR should never be created if the VRG is being deleted.  So as far as I see, there is no issue here.

BenamarMk replied:
Oh, I think it is [this line](https://github.com/RamenDR/ramen/blob/400fddb4c913a1e9f53af02d51a71fe5fa51ce53/controllers/volumereplicationgroup_controller.go#L661). Ignore my comment then.

ShyamsundarR replied:
:) ack! I was just creating this comment... processForDeletion -> reconcileVRsForDeletion -> reconcileVRForDeletion -> processVRAsPrimary/Secondary -> createOrUpdateVR...

BenamarMk reviewed (approved):
LGTM. I only have one question.

ShyamsundarR commented on controllers/volumereplicationgroup_controller.go:
```diff
@@ -617,14 +617,10 @@ func (v *VRGInstance) reconcileVRsForDeletion() bool {
-			continue
-		}
+		// Skip unprotected PVCs during delete
+		if pvc.Annotations[pvcVRAnnotationProtectedKey] != pvcVRAnnotationProtectedValue {
```
Having some doubts on what to check to check to ignore a PVC, I think it should be the finalizer and not the annotation here. (@BenamarMk @raghavendrabhat )

BenamarMk replied:
So the Annotation guarantees that even the PV metadata has been uploaded to s3.  Whereas the Finalizer doesn't guarantee that. However, it's error-prone if we use the annotation to check for VRProtection.  Because, if we fail to upload to s3 store, then the VRAnnotation will not be set. Hence, we skip reconciling VR for deletion.  so I think it is a good catch @ShyamsundarR.

ShyamsundarR commented:
@BenamarMk and @raghavendrabhat This PR is now dependent on #218 

The change to logic here is to keep processing PVCs that are marked with finalizers, but ensure VR is in the desired state before deletion. To prevent resource leaks in the storage system.

Also addressed the S3 deletion issue here, but we can decide if we want that in #218 and removed from here.

BenamarMk commented on controllers/volumereplicationgroup_controller.go:
```diff
@@ -61,7 +61,7 @@ type PVUploader interface {
 
 type PVDeleter interface {
-	DeletePV(v interface{}, s3ProfileName string, pvc *corev1.PersistentVolumeClaim) error
+	DeletePV(v interface{}, s3ProfileName string) error
```
(nit) DeletePV name should probably change to DeletePVs now that it is deleting a bunch.

ShyamsundarR replied:
True.. will make this change.

ShyamsundarR replied:
done!

raghavendrabhat replied:
Should there be a function to handle deletion of an individual PV from S3 store?

BenamarMk replied:
I am not sure if necessary, but all the plumbing is there, so if needed, we'll add one.

raghavendrabhat replied:
ACK

BenamarMk reviewed (approved):
LGTM. We can merge it as is.  I can simply close #218 once this is merged.

ShyamsundarR commented:
> LGTM. We can merge it as is. I can simply close #218 once this is merged.

Hmmm. that works as well.. as we have commit history with the right names.

raghavendrabhat reviewed (approved):
LGTM
